index.js

```diff
@@ -1,133 +1,69 @@
-    compilerOpts.contracts_build_directory = buildPath
-    compilerOpts.logger = Logger
-    compilerOpts.all = false
-
-    var provisionOpts = {}
-    provisionOpts.provider = new Web3.providers.HttpProvider(buildOpts.web3_rpc_uri)
-    provisionOpts.contracts_build_directory = buildPath
-
-    TruffleContractCompiler.compile(compilerOpts, function (err, contracts) {
-      if (err) {
-        Logger.error(err)
-        return compilationFinished(err, null)
-      }
-
-      isCompilingContracts = false
-      Logger.log('COMPILATION FINISHED')
-      Logger.log('RUNNING MIGRATIONS')
-
-      var migrationOpts = compilerOpts
-      migrationOpts.migrations_directory = buildOpts.migrations_directory
-      migrationOpts.contracts_build_directory = buildPath
-      migrationOpts.provider = provisionOpts.provider
-      migrationOpts.logger = Logger
-      migrationOpts.reset = true                                 // Force the migrations to re-run
-
-      // Once all of the contracts have been compiled, we know we can immediately
-      // try to run the migrations safely.
-      TruffleContractMigrator.run(migrationOpts, function (err, result) {
-        if (err) {
-          Logger.error(err)
-          return compilationFinished(err, null)
-        }
-        console.log('here', result)
-        // Finally return the contract source we were originally asked for.
-        returnContractAsSource(compiledContractPath, compilationFinished)
-      })
-    })
+  },
+  'additionalProperties': false
+}
 
-    return
+module.exports = function (source, map, meta) {
+  let WebpackOptions = getOptions(this) || {}
+  validateOptions(schema, WebpackOptions, 'truffle-solidity-loader')
+
+  let buildOpts = genBuildOptions(WebpackOptions)
+  let migrationsDirectory = WebpackOptions.migrations_directory || `${buildOpts.working_directory}/migrations`
+  let contractsBuildDirectory = WebpackOptions.contracts_build_directory || `${buildOpts.working_directory}/build/contracts`
+  let contractName = parseContractName(this.resourcePath) // this.resourcePath will be the path to the .sol file
+  let contractJsonPath = path.resolve(buildOpts.contracts_build_directory, contractName + '.json')
+  // this.addDependency(contractJsonPath); // NOTE adding dependency causes this to run twice
+
+  if (this.debug) {
+    Logger.debugger(`this.resourcePath = ${this.resourcePath}`)
+    Logger.debugger(`contract Name = ${contractName}`)
+    Logger.debugger(`migrations Directory = ${migrationsDirectory}`)
+    Logger.debugger(`contracts Build Directory = ${contractsBuildDirectory}`)
+    Logger.debugger(`contract Json Path = ${contractJsonPath}`)
   }
 
-  if (compiledContractExists(compiledContractPath)) {
-    returnContractAsSource(compiledContractPath, compilationFinished)
-  } else {
-    waitForContractCompilation()
-  }
+  let callback = this.async()
+  truffleMigrator.run(buildOpts, function (err) {
+    if (err) {
+      this.emitError(err)
```

Comment:
@sogoiii -- Whenever there's an error while trying to deploy the contracts on the app start, I get `this.emitError()` is not a function error. Please see the stacktrace below --

```
[! TRUFFLE SOLIDITY ERROR] ProviderError: Could not connect to your Ethereum client. Please check that your Ethereum client:
    - is running
    - is accepting RPC connections (i.e., "--rpc" option is used in geth)
    - is accessible over the network
    - is properly configured in your Truffle configuration file (truffle.js)

/node_modules/web3/lib/web3/httpprovider.js:128
      callback(error, result);
      ^
TypeError: this.emitError is not a function
    at /node_modules/truffle-solidity-loader/index.js:66:12
```

Is this the intended behaviour or should we rather shift to `callback(err, null)`? Thanks!

Reply:
Hey @eswarasai, thanks for testing this out! I am in contact with the truffle team and this will be integrated into their mono repo. I don't think a timeline has been set in stone, but hopefully soon enough :)

In regard to your error, ill look into it in a couple of hours. Im going to guess the `this` context inside of `truffleMigrator.run` is not the webpack context `this`.

Reply:
@sogoiii -- Thanks for the quick response. That's great news. I hope that happens soon enough  though. Solidity contract hot reloading is the best feature to have and eases development a lot.

Yeah, probably something similar. I wasn't sure either and haven't actually debugged it yet. Let me know what you find and I'll see if I can be of any assistance here.

Once again, thanks a lot for upgrading the solidity loader to work with the latest truffle :)